### PR TITLE
feat(automation) : Add custom keyword.conf file

### PR DESCRIPTION
**Pull request by rajuljha**
## Description

Add custom keywords.conf file by setting env varaible in GH Actions or GitlabCI.

### Changes

* Add `--keyword-conf` CLI argument to fossologyscanner script.
* Script takes the file, validates it, and copies it to desired location checked by scanner.
* Add helper functions to check the validity of input file as the scanner breaks if invalid file is given.
* Modifies GH Actions workflow file to use qemu for cross platform support.

## How to test

* Take new workflow file for GH Actions. Previous works fine with GitalbCI
* Use [this](https://hub.docker.com/layers/rjknightmare/fo-ci-test/latest/images/sha256-ac72768b8c4383a4aca05a50cd53ae427db9645b42619ad82b53e38b3b876b38?context=repo) test docker image (`rjknightmare/fo-ci-test:latest`) for testing.
* Create a new `keyword.conf` file according to this [format](https://github.com/fossology/fossology/blob/master/src/copyright/agent/keyword.conf) in your workflow repo.
* Set up an environment variable called `KEYWORD_CONF_FILE_PATH` and set it's value to the path of the file in your repo (in which you have set up the workflow) and run the CI pipeline.

### Building image inside the GH Action itself.
Screenshot 2024-06-23 at 7 43 40 PM

## With qemu 😀   
Screenshot 2024-06-23 at 7 44 07 PM

## GitlabCI
Screenshot 2024-06-23 at 7 48 53 PM

CC @GMishx @shaheemazmalmmd @Kaushl2208 @avinal

### CODE REVIEW

File: utils/automation/FoScanner/Utils.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+      # Check for the SPACESALL definition on the first non-commented line
+      spacesall_pattern = re.compile(r'SPACESALL=\[\[:space:\]\]\*')
+      if not spacesall_pattern.match(non_commented_lines[0]):
+        return False, "First non-commented line must define SPACESALL as [[:space:]]*"
+
+      # Check each subsequent non-commented line for the keyword pattern
+      keyword_pattern = re.compile(r'keyword=__keyword__\|?\(.*\)')
+      for line in non_commented_lines[1:]:
+        if not keyword_pattern.match(line):
+          return False, f"Invalid keyword pattern: {line}"
```
Comment (GMishx) [inline utils/automation/FoScanner/Utils.py:43]: There is no strict requirement to define `SPACESALL` in the conf file. All what it should contain is at least a `keyword=` line.

And the special symbol `__` must always occur in pair (if it exists).

And the non-comment lines must contain only 1 `=`

File: utils/automation/FoScanner/Utils.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+      return True, "Valid keyword.conf file"
+    
+    except FileNotFoundError:
+      return False, "File not found"
+    except Exception as e:
+      return False, str(e)
+
+def copy_keyword_file_to_destination(source_path:str, destination_path:str) -> None:
+    """
+    Copy keyword file to destination
```
Comment (GMishx) [inline utils/automation/FoScanner/Utils.py:54]: Please add that this function also makes the destination path.

File: utils/automation/FoScanner/CliOptions.py [hunk 3]
```diff
@@ -40,6 +40,7 @@ class CliOptions(object):
   keyword: bool = False
   repo: bool = False
   diff_dir: str = os.getcwd()
+  keyword_conf_file_path : str = ''
```
Comment (GMishx) [inline utils/automation/FoScanner/CliOptions.py:43]: Please document the class variable in the class documentation as `:ivar:`

**Review by GMishx — CHANGES_REQUESTED** [review]

File: utils/automation/FoScanner/Utils.py (hunk 2 above)
Comment (rajuljha) [reply utils/automation/FoScanner/Utils.py:54]: Corrected!

File: utils/automation/FoScanner/CliOptions.py (hunk 3 above)
Comment (rajuljha) [reply utils/automation/FoScanner/CliOptions.py:43]: Corrected!

File: utils/automation/FoScanner/Utils.py (hunk 1 above)
Comment (rajuljha) [reply utils/automation/FoScanner/Utils.py]: Modified it accordingly.
Please let me know if I missed something.

**Comment by GMishx** [comment]
@rajuljha please check the pipeline, there are missing license tags in utils file.

**Comment by rajuljha** [comment]
> @rajuljha please check the pipeline, there are missing license tags in utils file.

I have corrected it.

**Comment by Kaushl2208** [comment]
@GMishx , I have tested this for GH Actions. Looks good!!
image

**Review by GMishx — CHANGES_REQUESTED** [review]
Minor change needed.

**Review by GMishx — APPROVED** [review]
Changes looks good.

**Pull request merged**